content/docs/instrumenting/clientlib_guidelines.md

```diff
@@ -0,0 +1,250 @@
+
+In addition ENCOURAGED means that a feature is desirable for a library to have, but it’s okay if it’s not present. In other words, a nice to have.
+
+Things to keep in mind
+
+* Take advantage of each language’s features
+
+* The common use cases should be easy
+
+* The correct way to do something should be the easy way
+
+* More complex use cases should be possible
+
+The common use cases are (in order):
+
+* No-label counters spread liberally around libraries/applications
+
+* Timing functions/blocks of code in Summaries/Histograms
+
+* Gauges to track current states of things (and their limits)
+
+* Monitoring of batch jobs
+
+## Overall structure
+
+Clients MUST be written to be callback based internally. Clients SHOULD generally follow the structure described here.
+
+The key class is the Collector. This has a method (typically called ‘collect’) that returns zero or more metrics and their samples. Collectors get registered with a CollectorRegistry. Data is exposed by passing a CollectorRegistry to a class/method/function "bridge", which returns the metrics in a format Prometheus supports. Every time the CollectorRegistry is scraped it must callback to each of the Collectors’ collect method.
+
+The interface most users interact with are the Counter, Gauge, Summary, and Histogram Collectors. These represent a single metric, and should cover the vast majority of use cases where a user is instrumenting their own code.
+
+More advanced uses cases (such as proxying from another monitoring/instrumentation system) require writing a custom Collector. Someone may also want to write a "bridge" that takes a CollectorRegistry and produces data in a format a different monitoring/instrumentation system understands, allowing users to only have to think about one instrumentation system.
+
+CollectorRegistry SHOULD offer `register()`/`unregister()` functions, and a Collector SHOULD be allowed to be registered to multiple CollectorRegistrys.
+
+Client libraries MUST be thread safe.
+
+For non-OO languages such as C, client libraries should follow the spirit of this structure as much as is practical.
+
+### Naming
+
+Client libraries SHOULD follow function/method/class names mentioned in this document, keeping in mind the naming conventions of the language they’re working in. For example, `set_to_current_time()` is good for a method name Python, but `SetToCurrentTime()` is better in Go and `setToCurrentTime()` is the convention in Java. Where names differ for technical reasons (e.g. not allowing function overloading), documentation/help strings SHOULD point users towards the other names.
+
+Libraries MUST NOT offer functions/methods/classes with the same or similar names to ones given here, but with different semantics.
+
+## Standard metrics types
+
+Counter, Gauge, Summary and Histogram are the primary interface by users.
```

Comment:
Maybe `The Counter, Gauge, Summary and Histogram [Metric types](/docs/concepts/metric_types/) are the primary interface by users.`